src/stdlib/table.rs

```diff
@@ -731,3 +759,130 @@ fn array_insert_shift<'gc>(
     let length = table.length() as usize;
     (inner(table, length, key, value), length)
 }
+
+#[derive(Debug, Copy, Clone, Error)]
+pub enum TableConcatError {
+    #[error("resulting string too long in table.concat")]
+    Overflow,
+}
+
+fn concat_impl<'gc>(ctx: Context<'gc>, sep: Option<Value<'gc>>) -> BoxSequence<'gc> {
+    async_sequence(&ctx, |locals, mut seq| {
+        let sep = sep.map(|s| locals.stash(&ctx, s));
+        async move {
+            let args = seq.try_enter(|ctx, locals, _, mut stack| {
+                let sep_str = match sep
+                    .as_ref()
+                    .map(|s| locals.fetch(s))
+                    .map(|s| s.into_string(ctx))
+                {
+                    Some(Some(s)) => Some(s),
+                    Some(None) => return Ok(stack.len()),
+                    None => None,
+                };
+
+                let mut len = 0;
+                // Since we have to make two passes, might as well estimate the length
+                for value in &stack[..] {
+                    match value {
+                        Value::Integer(i) => len += i.ilog10() as usize + i.is_negative() as usize,
+                        Value::Number(_n) => len += 10,
+                        Value::String(s) => len += s.as_bytes().len(),
+                        _ => return Ok(stack.len()),
+                    }
+                }
+
+                let sep_len = sep_str.map(|s| s.len() as usize).unwrap_or(0);
+                let total_len = stack
+                    .len()
+                    .saturating_sub(1)
+                    .checked_mul(sep_len)
+                    .ok_or(TableConcatError::Overflow)?
+                    .checked_add(len)
+                    .ok_or(TableConcatError::Overflow)?;
+
+                // Should this be allocated in-place in the GC heap?
```

Comment:
Unfortunately this is impossible right now, this is an open design question in `gc-arena`. `gc-arena` currently relies on size being a static property of an allocated type. There's a hack (partially) around this in piccolo's `String` but it's horrible.